dev/buildtool/inspection_commands.py

```diff
@@ -12,100 +12,1154 @@
-        prefix=self.__bintray_url_prefix, decorator=decorator,
-        name=name, version=name_version)
-    logging.debug('Waiting for %s', url)
-    wait_until = time.time() + self.options.timeout
-    counted = False
+      self.__basic_auth = None
+
+  def fetch_bintray_url(self, bintray_url):
+    request = urllib2.Request(bintray_url)
+    if self.__basic_auth:
+      request.add_header('Authorization', self.__basic_auth)
+    try:
+      response = urllib2.urlopen(request)
+      headers = response.info()
+      payload = response.read()
+      content = json.JSONDecoder(encoding='utf-8').decode(payload)
+    except urllib2.HTTPError as ex:
+      raise_and_log_error(
+          ResponseError('Bintray failure: {}'.format(ex),
+                        server='bintray.api'),
+          'Failed on url=%s: %s' % (bintray_url, ex.message))
+    except Exception as ex:
+      raise
+    return headers, content
+
+  def list_bintray_packages(self, subject_repo):
+    path = 'repos/%s/packages' % subject_repo
+    base_url = 'https://api.bintray.com/' + path
+    result = []
     while True:
-      try:
-        urllib2.urlopen(url)
-        logging.debug('%s available', name_version)
-        return name_version
-      except urllib2.HTTPError as ex:
-        if not counted:
-          self.metrics.inc_counter(
-              'BintrayNotReady', {'repository': repository.name},
-              'Bintray not showing debian package that should be there.')
-          logging.warning('%s is not yet available on bintray: %s',
-                          name_version, ex)
-          counted = True
-
-        remaining = wait_until - time.time()
-        if remaining <= 0:
-          raise_and_log_error(
-              ResponseError(
-                  'Bintray not showing %s - giving up' % name_version,
-                  server='bintray'))
-        time.sleep(min(5, remaining))
-
-
-class WaitForDebiansFactory(RepositoryCommandFactory):
+      url = base_url + '?start_pos=%d' % len(result)
+      headers, content = self.fetch_bintray_url(url)
+      # logging.debug('Bintray responded with headers\n%s', headers)
+      total = headers.get('X-RangeLimit-Total', 0)
+      result.extend(['%s/%s' % (subject_repo, entry['name'])
+                     for entry in content])
+      if len(result) >= total:
+        break
+    return result
+
+  def query_bintray_package_versions(self, package_path):
+    path = 'packages/' + package_path
+    url = 'https://api.bintray.com/' + path
+    headers, content = self.fetch_bintray_url(url)
+    # logging.debug('Bintray responded with headers\n%s', headers)
+    package_name = package_path[package_path.rfind('/') + 1:]
+    return (package_name, content['versions'])
+
+  def difference(self, versions, target):
+    missing = []
+    for version in versions:
+      if not version in target:
+        missing.append(version)
+    return missing
+
+  def find_missing_jar_versions(self, jar_map, debian_map):
```

Comment:
We dont execute jars. We dont really need to know the jar is missing, but I thought it would be informative in determining integrity and deciding what can be deleted. If some services "expect" jars and they are not present, then those services might be desired to discard. The jars arent part of the runtime, but could be part of client/interoperational compatability where they are used in other service builds, meaning you would not be able to build against that version.